### ISSUE TITLE
Suggestion to add a seed()

### ISSUE THREAD

**Issue by CMCDragonkai** [robmorgan/phinx#172]
A suggestion to add a seed() function that is defined inside any migration classes. If it exists, Phinx will run the seed() only after it runs up().

I often need this to seed tables with initial data, which goes along with migration.

**Comment by alexbowers-tecmark** [comment]
That seems suitable to me.

**Comment by gmauricio** [comment]
is there someone working on this? @todiadiyatmo, @robmorgan ? 
If not I'd be interested in doing it.

**Comment by robmorgan** [comment]
I'm not working on this at the moment, but I'd happily accept a PR. Also I'd prefer we develop it with only one command. i.e: `phinx seed`

**Comment by gmauricio** [comment]
Hey @robmorgan , please check #457, there I added an insert function to the adapter interface, also added it to Table. The idea is to implement seeder classes as in Laravel, and then they would have access to the same Table as in migrations.   Like this:

```
class DatabaseSeeder extends SeederInterface
{
    function run()
    {
        $user = $this->table('user');
        
        $data = array(
            array("email" => "foo@bar.com")
        );

        $user->insert($data)->save();
    }
}
```

So, in the PR, I already enabled that for Table. 

Now I'm not sure if we should separate the interfaces for the Table used in seeders, and the Table used in migrations, to avoid being able to for example do changes to the schema in a seeder, or maybe add data in a migration. My opinion is that there's no need to separate them, it would be up to the developer to use it correctly.

I'll implement the seed command, and the Seeder base class after getting #457 accepted.

**Comment by alexbowers-tecmark** [comment]
How does this handle duplicate keys, and other exceptions. Are they all thrown to the user migrating and seeding, or are they just ignored?

Also, how is the wrong data type handled. Since MySQL (and I assume other drivers) would throw an exception; does this stop execution and clear the data that has already been inserted, or is it non-transitional.


----------------------

Alex Bowers
Backend Developer

tecmark
Manchester & London
Tel:  0845 680 1146
Fax:  0845 680 1316
Email:  alex.bowers@tecmark.co.uk<mailto:alex.bowers@tecmark.co.uk>
Web:  www.tecmark.co.uk<http://www.tecmark.co.uk/>

111 Piccadilly, Manchester M1 2HY<http://maps.google.com/maps?q=111%20Piccadilly%2C%20Manchester%20M1%202HY>

City Point, 1 Ropemaker Street, London EC2Y 9HT<http://maps.google.com/maps?q=City%20Point%2C%201%20Ropemaker%20Street%2C%20London%20EC2Y%209HT>

**Comment by gmauricio** [comment]
* Currently not handling duplicate keys.
* Currently, wrong data types are not handled

Exceptions will be thrown to the user migrating or seeding.
I assume you meant "not-transactional". As I saw in the code, transactions are already managed on Phinx\Migration\Manager\Environment, which wraps the actual execution of any adapter function in the context of a migration execution. For seeding it would be done in a similar way.

Right now, I'm reviewing the logic to decide which kind of checks to add or exceptions to handle.

**Comment by alexbowers-tecmark** [comment]
Any update on this?

**Comment by gmauricio** [comment]
Hey @alexbowers-tecmark , I didn't have much time to work on this until now.
I just implemented some changes according to @shadowhand feedback.

About validations, I'm not sure what to implement, if having duplicate keys, or wrong data types, the migration would just fail at the db level, and I think that's ok, what do you think @shadowhand ?

**Comment by BlakeGardner** [comment]
@gmauricio I think just letting the seeder fail and throw an exception is just fine. I already have a hacked together seeder system based on symfony console and i just get a big red exception in my terminal window which works.

**Comment by alexbowers-tecmark** [comment]
I have to agree on this. If it throws an exception, that is fine. But the seeder should come with a ‘refresh’ method, which wipes all the tables before it inserts into them to cover the lack of a transaction

**Comment by mrova** [comment]
+1 for the seed command

**Comment by alekseyp** [comment]
:+1: for seed

**Comment by cezarykluczynski** [comment]
:+1: I'd love to see this happen. Installing Laravel just to have decent migrations seems a little overhead.

**Comment by BlazOrazem** [comment]
:+1: for seed command

**Comment by BlakeGardner** [comment]
I have multiple code bases which rely on the same database schema so we choose phinx because it is framework agnostic. At this point i am considering just doing a bare Laravel install so i can get seeder support.

**Comment by robmorgan** [comment]
Good news guys, I started coding this feature last week. I tried to finish it before my talk at PHP Tour Luxembourg, but didn't in time! Stay tuned though, will be released in 0.5.0 (next release).

**Comment by gmauricio** [comment]
Hey @robmorgan , some time ago I started this (https://github.com/robmorgan/phinx/compare/0.4.x-dev...gmauricio:seeder-command), but didn't get the time to properly test it. Maybe you can find something useful there. Also, I have some time now, let me know if I can be of any help.

**Comment by rockpunk** [comment]
Hi guys, this isn't _exactly_ the requested feature for a `seed()` method as some of you have proposed, but I have recently authored pull request #610 which modifies phinx to add a very simple seed data and structure dump. It may be sufficient for your needs.

It's unfortunately failing some automated travis builds due to a travis config, but it could use a look through from the community.

**Comment by thomasbiddle** [comment]
+1 for seed command. Looking forward to 0.5.0!

**Comment by robmorgan** [comment]
@gmauricio thanks mate it was actually pretty similar to my implementation so I borrowed a few bits.

**Comment by gmauricio** [comment]
@robmorgan , good to hear :smile:

**Comment by 2upmedia** [comment]
:clap::clap::clap:

**Comment by dereuromark** [comment]
New features and discussions around it should be a new ticket, not opening up a 3 year old one.

**Comment by OwenMelbz** [comment]
@dereuromark it wasn't an intention to request a new feature - more of asking a question surrounding this existing topic?

Thanks

**Comment by robmorgan** [comment]
@OwenMelbz hmmm not sure if this is a good idea, but I can't think of anything bad right at the minute. The only problem is, you'd have to handle the data cleanup in the `down()` method, so maybe it should be forbidden inside a `change()` method.

Anyway, I agree with @dereuromark - open a new ticket and lets discuss it there.
